### PR TITLE
lora: Add support for fine-grained STM32WL power amplifier configuration

### LINKED ISSUE #48511: STM32WL LoRaWAN pinout incomplete / RFO pin not selectable
**Is your enhancement proposal related to a problem? Please describe.**

The STM32WL chip series has on chip LoRa and provides two input and two output pins for this. The low power output pin (RFO_LP) and the high power output pin (RFO_HP) are currently not choosable. By default the RFO_HP pin is used.

The limitation was found because of the following problem:
A custom board with STM32WLE5CC, which is optimized to only use the RFO_LP pin was configured with two different firmwares.
The first firmware is a bare metal version configured with STM32CubeMX and the second one is the lorawan class a sample from zephyr. The bare metal firmware reached and rssi of around -70 and the zephyr sample only of -110. After connecting the RFO_LP and the RFO_HP pin the zephyr sample reached an rssi of around -60.

**Describe the solution you'd like**
An optimal solution would be an option in the devicetree to select the preferred RFO pin.

**Additional context**
From the Datasheet of STM32WLE5CC
image

### PR DESCRIPTION
**Pull request by knthm**
### Summary

This PR adds the ability to select between the low-power `RFO_LP` and high-power `RFO_HP` LoRa radio power amplifier output, as well as setting the maximum design Tx output via a new devicetree properties on STM32WL-based boards.
This allows better matching of the power amplifier output with boards' RF impedance matching network.

Previously only the high-power `RFO_HP` output was supported in the highest design power output (+22 dBm), as the same PA config register is used to differentiate between the discrete SX1261 and SX1262 chips in LoRaMac-node.

I've locally tested this change on a Olimex STM32WL Devkit.

Closes #48511

### Change Description

Add STM32WL-specific `sx126x_set_tx_params` function based on the STM32CubeWL modifications to LoRaMac-node.

~~Add `tx-power-low` devicetree binding to `st,stm32wl-subghz-radio` for toggling the `RFO_LP` output functionality provided in this commit.~~
**Add the `power-amplifier-output` devicetree property to `st,stm32wl-subghz-radio` for selecting between the RFO_LP and RFO_HP output configurations provided by the above mentioned function.**

Also add the `rfo-lp-max-power` and `rfo-hp-max-power` DT properties for defining the maximum design power of the respective outputs' matching networks.

~~Set `RFO_LP` output as default for the BB-STM32WL module.~~
**Set the appropriate output and max-power values for all affected boards and modules.**
### Impact and future work

I consider this method of least impact for adding this feature and aligning the behavior of the Zephyr LoRa driver for the STM32WL family with that provided by the STM32CubeWL driver. (Versus the alternative of adding the STM32CubeWL LoRaMac-node modifications as its own Zephyr module with accompanying glue code in `sx126x.c`)

This PR may improve `RFO_LP` Tx performance, as it uses the higher duty cycle configured in the STM32CubeWL driver for the +15 dBm output mode. It should also guarantee better device longevity by additionally enforcing overcurrent protection for the power amplifier.

This PR also paves the way for more advanced PA output selection. A `tx-power-gpios` property might later be added for defining GPIOs to control an RF switch to toggle `RFO_LP` and `RFO_HP` on hardware supporting this such as the Nucleo WL55JC.

There are still some differences in hardware configuration to STM32CubeWL that the Zephyr LoRa driver is lacking, such as not setting the highest Rx gain possible in `SX126xSetRxBoosted`, and the LoRaMac-node code assuming a default antenna gain of 2.15 dBi, but I will address these in separate PR(s).

@martinjaeger I'm not sure how you've worked around this issue on your hardware, but if you've re-soldered the 0R resistor to enable the `RFO_HP` output path, you'll need to delete the newly added property in your overlay(s). Just a friendly heads up :smiley:

### References

See chapter RM0461 chapter 4.8.4., Set_PaConfig() command, and AN5457 chapter 5.1.2.

1. [RM0461 Reference manual - STM32WLEx advanced Arm-based 32-bit MCUs with sub-GHz radio solution](https://www.st.com/resource/en/reference_manual/rm0461-stm32wlex-advanced-armbased-32bit-mcus-with-subghz-radio-solution-stmicroelectronics.pdf)
2. [AN5457 Application note - RF matching network design guide for STM32WL Series](https://www.st.com/resource/en/application_note/an5457-rf-matching-network-design-guide-for-stm32wl-series-stmicroelectronics.pdf)
3. [STM32CubeWL SubGHz_PHY driver](https://github.com/STMicroelectronics/STM32CubeWL/blob/main/Middlewares/Third_Party/SubGHz_Phy/stm32_radio_driver/radio_driver.c#L620-L692)

### CODE REVIEW

**Comment by KloudJack** [comment]
- Amend formatting to comply with coding guidelines

**Comment by KloudJack** [comment]
- Amend formatting to comply with coding style guidelines

**Comment by erwango** [comment]
Direction looks ok to me but I'd like actual and active WL users to provide feedback first

**Comment by erwango** [comment]
@danielSchaeble, @aurel32

**Comment by danielSchaeble** [comment]
I tested it on custom hardware with STM32WLE5CC and the RFO_LP pin attached. Works fine with the new devicetree flag activated. Thanks! :)

**Review by martinjaeger — COMMENTED** [review]
Thanks a lot @KloudJack.

I guess I haven't realized this issue as my LoRa devices based on this board are very close to the gateways in all my current applications.

I'll try to test it out next week. Some minor comments below.

File: drivers/lora/sx126x_common.h [hunk 1]
```diff
@@ -64,6 +65,8 @@ void sx126x_dio1_irq_enable(struct sx126x_data *dev_data);
 
 void sx126x_dio1_irq_disable(struct sx126x_data *dev_data);
 
+void sx126x_set_tx_params(int8_t power, RadioRampTimes_t rampTime);
```
[the remark this answers is not in the document]
Comment (KloudJack) [reply drivers/lora/sx126x_common.h:68]: Good catch! I'll amend the original commit.

File: drivers/lora/sx126x_standalone.c [hunk 2]
```diff
@@ -61,6 +61,11 @@ static void sx126x_dio1_irq_callback(const struct device *dev,
 	}
 }
 
+void sx126x_set_tx_params(int8_t power, RadioRampTimes_t rampTime)
+{
+	SX126xSetTxParams(power, RADIO_RAMP_40_US);
```
[the remark this answers is not in the document]
Comment (KloudJack) [reply drivers/lora/sx126x_standalone.c:66]: Nope, just a late-night snafu :)

File: drivers/lora/sx126x_stm32wl.c [hunk 3]
```diff
@@ -16,6 +16,8 @@
 #include <zephyr/irq.h>
 LOG_MODULE_DECLARE(sx126x, CONFIG_LORA_LOG_LEVEL);
 
+static const bool subghz_rfo_lp = DT_INST_NODE_HAS_PROP(0, tx_power_low);
```
[the remark this answers is not in the document]
Comment (mniestroj) [reply drivers/lora/sx126x_stm32wl.c:19]: Since this variable represents property that for `st,stm32wl-subghz-radio.yaml` (i.e. stm32wl specific), I would rather leave it here as is.

[the remark this answers is not in the document]
Comment (KloudJack) [reply drivers/lora/sx126x_stm32wl.c:19]: I agree with your argument about consistency, but I'm kind of split due to the added code verbosity.

I guess the other conditional  `sx126x_config` members used in `sx126x_standalone.c` might technically also be used in the main driver or other future variants, while the `RFO_LP` property only ever really applies to the Tx parameters on the STM32WL.

Would it make sense to muddy the general `sx126x` device config with a member and `#defines` that'd wouldn't have any use outside the small scope where they're currently used? Either way I don't mind adding it or leaving it, maybe somebody else has some feedback on this.

**Comment by KloudJack** [comment]
- Amend commit with review suggestions

**Comment by aurel32** [comment]
Thanks for the PR. I haven't look at all the details, but overall the approach looks fine to me. I just wonder if the device tree entry shouldn't be an enum with the options "hp" or "lp". Sure it means changing the dts of existing boards, but it's probably less error prone in the long term. It means users really have to select the right output instead of possibly finding that a board doesn't work.

I'll try to review the PR more in details in the next days.

**Comment by mniestroj** [comment]
> I just wonder if the device tree entry shouldn't be an enum with the options "hp" or "lp".

I like the idea! I was also thinking about changing the name of this property so it would better describe that it is about selecting `RFO_LP`. We already have following properties for `sx127x`:
- `power-amplifier-output = "rfo";` (which is an enum)
- `rfo-enable-gpios = <&gpios ...>;`

1) so I would add something like:
- `rfo = "hp";`
- `rfo = "lp";`

2) or we could use `power-amplifier-output` just like with `sx127x`:
- `power-amplifier-output = "rfo-hp";`
- `power-amplifier-output = "rfo-lp";`
with following bindings:
```
  power-amplifier-output:
    type: string
    required: true
    description: |
      Selects power amplifier output pin.
    enum:
      - "rfo-lp"
      - "rfo-hp"
```

I think the latter makes more sense and will make it consistent across sx12xx families.

**Comment by KloudJack** [comment]
I originally drafted this PR with an enum as well, which would get overridden in `sx126x_set_tx_params` if `DT_INST_NODE_HAS_PROP(0, tx-power-gpios)` evaluated as true.

After some consideration I thought it'd be more likely to be get accepted if it didn't require adding a property to several boards that work with the default `RFO_HP` output exclusively anyway. AFAICT only the BB-STM32WL module doesn't work out of the box without the property selecting `RFO_LP`.

> or we could use power-amplifier-output just like with sx127x:

I do like the idea of matching the structure and behavior of the SX127x driver, so if nobody objects I'll edit the commit to use the original enum property and logic.

Excluding the GPIO logic though as that's not as cleanly implemented considering how some RF switches are wired.

*See the Nucleo WL55JC schematics, it's not as simple as having a dedicated GPIO line for toggling between `RFO_LP` and `RFO_HP`: the existing `FE_CTRL1` and `FE_CTRL2` lines which are individually used for switching the antenna between Rx and Tx are used for a third state with both lines logic high to toggle the `RFO_LP` output. Adding support for that would require changes to the existing Tx/Rx switch code in `SX126xSetOperatingMode` in the general SX126x driver which I wanted to avoid.*

**Comment by KloudJack** [comment]
Amended commit to employ enum property suggested above.

The latest commit now specifically adds the `RFO_HP` output selection to the Seeed Studio LoRa-E5 module and Nucleo WL55JC board (for which the `RFO_HP` RF switch output is selected inherently by the current driver behavior in switching the Rx/Tx output GPIOs through `FE_CTRL1` and `FE_CTRL2`. See my previous comment and [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/lora/sx126x.c#L294-L295)).

I chose not to define a default `RFO_LP`/`RFO_HP` output in `stm32wl.dtsi`, as this will prompt people with custom/out-of-tree boards to update their devicetree definitions when they run into this, rather than silently enforcing a default which may be incorrect for their use case.

Thanks for the review and your feedback on this @mniestroj @aurel32! :)

File: drivers/lora/sx126x_stm32wl.c [hunk 4]
```diff
@@ -45,6 +45,53 @@ void sx126x_dio1_irq_disable(struct sx126x_data *dev_data)
 	irq_disable(DT_INST_IRQN(0));
 }
 
+void sx126x_set_tx_params(int8_t power, RadioRampTimes_t ramp_time)
+{
+	uint8_t buf[2];
+
+	if (DT_INST_ENUM_IDX(0, power_amplifier_output) == 0) {
```
Comment (JordanYates) [inline drivers/lora/sx126x_stm32wl.c:52]: This is my only concern with the implementation, it is a brittle way of checking the enum value.
I have opened #56874 which will hopefully allow writing this in a clearer fashion.
`if (DT_INST_ENUM_HAS_VALUE(0, power_amplifier_output, rfo_lp))`

**Review by JordanYates — CHANGES_REQUESTED** [review]
No concerns with the approach taken, I would just like to see the linked PR merged first which would allow making the logic much clearer.

File: drivers/lora/sx126x_stm32wl.c (hunk 4 above)
Comment (KloudJack) [reply drivers/lora/sx126x_stm32wl.c:52]: I generally agree with your suggestion and associated PR, but the usage here IMHO is robust.

1. A DTS enum will always have at least one value which will match the first conditional outcome. For any other value, the alternative code path acts as a catch-all mimicking the behavior before this PR.
2.  The added property is marked `required`, so any DTS configurations activating this portion of the SX126x driver with `CONFIG_LORA_STM32WL_SUBGHZ_RADIO` through `DT_HAS_ST_STM32WL_SUBGHZ_RADIO_ENABLED` will fail the DTC process, should the property be missing or its value or not match one of the two defined values.
3. From a quick grep I see at least 9 other comparable uses of `DT_ENUM_IDX` and its variations throughout the Zephyr codebase. I think those and this PR's use of which should be addressed separately.

Comment (JordanYates) [reply drivers/lora/sx126x_stm32wl.c:52]: > I generally agree with your suggestion and associated PR, but the usage here IMHO is robust.

I disagree that it can be considered robust when simply swapping the order of `rfo-lp` and `rfo-hp` in the binding file will invert the behaviour of the code.

Comment (KloudJack) [reply drivers/lora/sx126x_stm32wl.c:52]: Naturally, enum ordering is important. This PR adds the enum property and the logic handling it in a single commit, why would the order in the binding definition be subject to change without consideration for the code handling it?

Also, how is this different from the dozens of other uses of `DT_ENUM_IDX` and co. where the enum index is directly operated on, such as  non-successive integer enums, or enums used indirectly with a matching C-source enum? (Which with the same argument may also deviate in order)

I'm all for decreasing ambiguity in DTS enum value operations, but PRs shouldn't suddenly be blocked if it already occurs several times in the existing codebase.

In my experience, creating dependencies between functionally unrelated PRs unnecessarily increases reviewer spread and slows down integration agility.

Comment (JordanYates) [reply drivers/lora/sx126x_stm32wl.c:52]: > Naturally, enum ordering is important. This PR adds the enum property and the logic handling it in a single commit, why would the order in the binding definition be subject to change without consideration for the code handling it?

If enum values are important, they should be of the form: `enum { VAL_A = 1, VAL_B = 2 };`. The problem here is the complete lack of any link between how the enum is generated (binding file) and where it is used (the comparison). This is the exact reason why enum support was improved as a result of #21273.

> Also, how is this different from the dozens of other uses of DT_ENUM_IDX and co. where the enum index is directly operated on, such as non-successive integer enums, or enums used indirectly with a matching C-source enum? (Which with the same argument may also deviate in order)

Just because it hasn't been caught by other reviewers doesn't make the solution any more correct IMO.

> I'm all for decreasing ambiguity in DTS enum value operations, but PRs shouldn't suddenly be blocked if it already occurs several times in the existing codebase.

If it is time critical for this to be merged for some reason, there are other options that aren't quite as ergonomic but don't have the same problem.
```
enum {
   PA_RFO_LP,
   PA_RFO_HP
} dt_inst_power_mode = UTIL_CAT(PA_, DT_INST_STRING_UPPER_TOKEN(0, power_amplifier_output));
...
if (dt_inst_power_mode == PA_RFO_LP) {}
```

Comment (KloudJack) [reply drivers/lora/sx126x_stm32wl.c:52]: > If it is time critical for this to be merged for some reason, there are other options that aren't quite as ergonomic but don't have the same problem.

I just want to avoid it getting mothballed by dependency-chaining it to to an issue which might evoke longer discussion, when it solves a real problem with a hardware feature for a supported board that *seems* to be operational, but really is set to a non-functional and dangerous (for the longevity of the hardware) configuration.

Comment (mniestroj) [reply drivers/lora/sx126x_stm32wl.c:52]: I like the use of `DT_INST_STRING_UPPER_TOKEN()`, as shown above.

Comment (KloudJack) [reply drivers/lora/sx126x_stm32wl.c:52]: OK, after some consideration I think it's a good compromise.

**Comment by KloudJack** [comment]
Amend commit with C source enum from DT_INST_STRING_UPPER_TOKEN, as suggested by @JordanYates

**Review by mniestroj — DISMISSED** [review]

File: drivers/lora/sx126x_stm32wl.c [hunk 5]
```diff
@@ -45,6 +50,53 @@ void sx126x_dio1_irq_disable(struct sx126x_data *dev_data)
...
+		if (power <= 20 && power > 17) {
+			SX126xSetPaConfig(0x03, 0x05, 0x00, 0x01); /* +20 dBm */
+		} else if (power > 14) {
+			SX126xSetPaConfig(0x02, 0x03, 0x00, 0x01); /* +17 dBm */
+		} else if (power <= 14) {
+			SX126xSetPaConfig(0x02, 0x02, 0x00, 0x01); /* +14 dBm */
+		} else {
+			SX126xSetPaConfig(0x04, 0x07, 0x00, 0x01); /* +22 dBm */
+			power = (power > 22) ? 22 : power;
+		}
```
Comment (aurel32) [inline drivers/lora/sx126x_stm32wl.c:87]: This  code seems to be unreachable, given that one of the `power > 14` and `power <= 14` conditions is always true.

What about ordering the comparisons, by decreasing values:
```

if (power >= 20) {
... /* +22 dBm */
} else if (power > 17) {
... /* +20 dBm */
} else if (power > 14) {
... /* +17 dBm */
} else {
... /* +14 dBm */
}
```

**Review by aurel32 — CHANGES_REQUESTED** [review]

**Comment by aurel32** [comment]
> Add STM32WL-specific sx126x_set_tx_params function based on the STM32CubeWL modifications to LoRaMac-node.

After reading AN5457, I am a bit concern by that change. Especially the fact that the application notes says "_This is because the RF PA output impedance values change with the RF PA settings._".

The STM32CubeWL does not change the PA config based on the **requested TX power,** but rather on the **maximum design TX power**. I believe that most manufacturers determine the RF matching network based on the maximum TX power allowed by either the LP or HP outputs, so changing the PA config might actually increase the RF mismatch,  reducing the effective TX power.

File: drivers/lora/sx126x_stm32wl.c [hunk 6]
```diff
@@ -16,6 +16,11 @@
 #include <zephyr/irq.h>
 LOG_MODULE_DECLARE(sx126x, CONFIG_LORA_LOG_LEVEL);
 
+static const enum {
+	RFO_LP,
+	RFO_HP,
+} pa_output = DT_INST_STRING_UPPER_TOKEN(0, power_amplifier_output);
+
```
[the remark this answers is not in the document]
Comment (KloudJack) [reply drivers/lora/sx126x_stm32wl.c:23]: Indeed, me too! I didn't think to extract a DT enum string value directly with `DT_STRING_TOKEN`, in retrospect it almost seems obvious.

We can thank @JordanYates for this insight. :)

File: drivers/lora/sx126x_stm32wl.c (hunk 5 above)
Comment (KloudJack) [reply drivers/lora/sx126x_stm32wl.c:87]: Good catch, your proposed edit is more readable as well.

However, I realize based on your comment regarding AN5457 that the conditional logic here needs to be reworked. Will address that below.

**Comment by KloudJack** [comment]
> The STM32CubeWL does not change the PA config based on the **requested TX power,** but rather on the **maximum design TX power**. I believe that most manufacturers determine the RF matching network based on the maximum TX power allowed by either the LP or HP outputs, so changing the PA config might actually increase the RF mismatch, reducing the effective TX power.

Damn good find, I should've read the application note more thoroughly.

Since we don't set `max_power` anywhere, and the BB-STM32WL module I work with uses the same [matching and filtering network](https://www.olimex.com/Products/IoT/LoRa/BB-STM32WL/resources/BB-STM32WL_RevC.pdf) for both +14 dBm `RFO_LP` and +22 dBm `RFO_HP` outputs (based on the configuration of a 0R resistor), I'd falsely assumed that we could just use the passed `power` value which seems to be based on the max EIRP for the configured region.

I think this calls for adding a DT enum for each output to select the maximum design RF output power of a board, and a rework of the logic that configures the PA based on those values.

On a similar note, I've found another interesting tidbit when traversing the LoRaMac-node code: it assumes 2.15 dBi default antenna gain in all regions, since we don't set the `MIB` antenna gain parameter in the driver. Users with lower gain antennas will find themselves not reaching their region's max EIRP, and users with higher gain ones will possibly be exceeding it (providing high gain antennas are even legal in their regions, EU rules at least seems to stipulate a max gain of 2.15 dBi).

**Comment by KloudJack** [comment]
- Add new DT properties `rfo-lp-max-power` and `rfo-hp-max-power`
- Update PA config logic, now functionally matches STM32CubeWL SubGHz driver

File: dts/bindings/lora/st,stm32wl-subghz-radio.yaml [hunk 7]
```diff
@@ -12,3 +12,33 @@ properties:
...
+      Selects between the low- and high-power power amplifier output pin.
+    enum:
+      - "rfo-lp"
+      - "rfo-hp"
+
+  rfo-lp-max-power:
+    type: int
+    default: 15
+    description: |
+      Maximum design power for the board's RFO_LP output matching network.
```
Comment (aurel32) [inline dts/bindings/lora/st,stm32wl-subghz-radio.yaml:29]: You might want to mention AN5457, so that users know from where those values come from.

**Review by aurel32 — DISMISSED** [review]
Beside minor nitpicking, it looks all good to me. Thanks for the work

File: dts/bindings/lora/st,stm32wl-subghz-radio.yaml (hunk 7 above)
Comment (KloudJack) [reply dts/bindings/lora/st,stm32wl-subghz-radio.yaml:29]: Good idea, I'll amend the commit.

Comment (aurel32) [reply dts/bindings/lora/st,stm32wl-subghz-radio.yaml:29]: Thanks!

**Comment by KloudJack** [comment]
Amended commit with reference to AN5457 in DT binding property description.

Thanks again for the thorough review @aurel32! :smile:
The previous code would've likely set suboptimal PA config values for several board configurations.

Having PA configuration settings to complement matching networks highest Tx output makes perfect sense, now that it's tumbled around my brain for a few hours—seeing how impedance is a function of reactance and resistance, with resistance varying based on power amplifier output current.

File: boards/arm/nucleo_wl55jc/nucleo_wl55jc.dts [hunk 8]
```diff
@@ -169,6 +169,13 @@
 		rx-enable-gpios = <&gpioc 5 GPIO_ACTIVE_LOW>;
 		dio3-tcxo-voltage = <SX126X_DIO3_TCXO_1V7>;
 		tcxo-power-startup-delay-ms = <5>;
+		/* Hard-code the high-power output, which is selected
```
Comment (JordanYates) [inline boards/arm/nucleo_wl55jc/nucleo_wl55jc.dts:172]: I would reword this to make it clear that we are not hardcoding the HP output, but it is hardcoded because we are (ab)using the `tx/rx-enable-gpio` to control `FE_CTRL1/2`,
```
High-power output is selected as a consequence of using tx/rx-enable-gpio
to control FE_CTRL1 and FE_CTRL2. Low-power output would require both
FE_CTRL1 and FE_CTRL2 to be high, which is not currently supported by the
driver.
```
I would also add a `FE_CTRL1` and `FE_CTRL2` comment on the two GPIO's above to make it as clear as possible.

**Review by JordanYates — CHANGES_REQUESTED** [review]
Some comments on the documentation and devicetree.

File: dts/bindings/lora/st,stm32wl-subghz-radio.yaml [hunk 9]
```diff
@@ -12,3 +12,37 @@ properties:
...
+    required: true
+    description: |
+      Selects between the low- and high-power power amplifier output pin.
+    enum:
+      - "rfo-lp"
+      - "rfo-hp"
+
+  rfo-lp-max-power:
+    type: int
+    default: 15
```
[the remark this answers is not in the document]
Comment (martinjaeger) [reply dts/bindings/lora/st,stm32wl-subghz-radio.yaml:27]: I would agree.

The AN5457 states:
image

`0x04` corresponds to 14, so the 15 is not a safe default for all boards.

File: drivers/lora/sx126x_stm32wl.c [hunk 10]
```diff
@@ -45,6 +50,71 @@ void sx126x_dio1_irq_disable(struct sx126x_data *dev_data)
...
+		if (power > max_power) {
+			power = max_power;
+		}
+		if (max_power == 14) {
+			SX126xSetPaConfig(0x04, 0x00, 0x01, 0x01);
+			power = 14 - (max_power - power);
+		} else if (max_power == 10) {
+			SX126xSetPaConfig(0x01, 0x00, 0x01, 0x01);
+			power = 13 - (max_power - power);
+		} else { /* default +15 dBm */
```
Comment (martinjaeger) [inline drivers/lora/sx126x_stm32wl.c:69]: This should probably not be considered the default, see also other comment.

**Review by martinjaeger — COMMENTED** [review]
I have tested this PR today and got very nice results :)

Before: SNR 1.5 RSSI -110
After: SNR 8.25 RSSI -52

Will approve after the suggestions from @JordanYates are addressed.

File: boards/arm/nucleo_wl55jc/nucleo_wl55jc.dts (hunk 8 above)
Comment (KloudJack) [reply boards/arm/nucleo_wl55jc/nucleo_wl55jc.dts:172]: That wording is much clearer, will update it thanks.

File: dts/bindings/lora/st,stm32wl-subghz-radio.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (KloudJack) [reply dts/bindings/lora/st,stm32wl-subghz-radio.yaml:27]: > It is mandatory for default values to have an explanation under `description:`

Thanks for the heads up, will add to the description.

> I suspect these properties would be better off without `default` and adding `required: true`

That'd require defining both output `max-power` properties for boards that will often only have one output connected, leading to a superfluous and erroneous DTS property, which I think would be more confusing. The other alternative of using some weird array combinations instead of two enums, or an additional per-property disabled value would be even clunkier I think.

> `0x04` corresponds to 14, so the 15 is not a safe default for all boards.

15 is the default in the STM32CubeWL, but there's probably some additional safeguard logic preventing that setting on 433 MHz boards. At the very least the `*_MAXPOWER` defines are mandatory there to prevent accidentally setting +15 dBm for those.

I'll change the defaults to 14 to be on the safe side, good call :)

File: drivers/lora/sx126x_stm32wl.c (hunk 10 above)
Comment (KloudJack) [reply drivers/lora/sx126x_stm32wl.c:69]: Agreed, see other comment, will update.

**Comment by KloudJack** [comment]
Amended commits:

- Add explanation of default value to DT property binding description
- Reword comment explaining `RFO_HP` selection in Nucleo WL55JC board for clarity
- Set `RFO_LP` `max-power` default to +14 dBm
____

> I have tested this PR today and got very nice results :)

Awesome! Thanks so much for testing :heart: 

Appreciate the review and suggestions you two. :)

**Review by JordanYates — APPROVED** [review]

**Review by martinjaeger — APPROVED** [review]

File: dts/bindings/lora/st,stm32wl-subghz-radio.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (martinjaeger) [reply dts/bindings/lora/st,stm32wl-subghz-radio.yaml]: Makes sense. Thanks for the final update!

**Review by mniestroj — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]
